web/src/user_group_edit_members.ts

```diff
@@ -179,32 +181,112 @@ export function rerender_members_list({
     });
 }
 
+function generate_group_link(group: UserGroup): string {
+    if (group.is_system_group) {
+        const group_name = user_groups.get_display_group_name(group.name);
+        return `<a data-user-group-id="${group.id}" class="view_user_group">${group_name}</a>`;
+    }
+    return `<a data-user-group-id="${group.id}" class="view_user_group">${group.name}</a>`;
+}
+
+function generate_user_link(user: User): string {
+    return `<a data-user-id="${user.user_id}" class="view_user_profile">${Handlebars.Utils.escapeExpression(user.full_name)}</a>`;
+}
+
+function generate_members_added_success_messages(
+    newly_added_users?: User[],
+    newly_added_subgroups?: UserGroup[],
+    already_added_users?: User[],
+    already_added_subgroups?: UserGroup[],
+): {
+    newly_added_members_message: string;
+    already_added_members_message: string;
+} {
+    const new_user_links = newly_added_users?.map((user) => generate_user_link(user)) ?? [];
+    const new_group_links = newly_added_subgroups?.map((group) => generate_group_link(group)) ?? [];
+    const old_user_links = already_added_users?.map((user) => generate_user_link(user)) ?? [];
+    const old_group_links =
+        already_added_subgroups?.map((group) => generate_group_link(group)) ?? [];
+
+    const newly_added_members_message = util.format_array_as_list_with_conjunction(
+        [...new_user_links, ...new_group_links],
+        "long",
+    );
+    const already_added_members_message = util.format_array_as_list_with_conjunction(
+        [...old_user_links, ...old_group_links],
+        "long",
+    );
+    return {
+        newly_added_members_message,
+        already_added_members_message,
+    };
+}
+
 function show_user_group_membership_request_result({
-    message,
+    error_message,
     add_class,
     remove_class,
     already_added_users,
+    newly_added_users,
+    newly_added_subgroups,
     ignored_deactivated_users,
     already_added_subgroups,
     ignored_deactivated_groups,
 }: {
-    message: string;
+    error_message?: string;
     add_class: string;
     remove_class: string;
+    newly_added_users?: User[];
+    newly_added_subgroups?: UserGroup[];
     already_added_users?: User[];
     ignored_deactivated_users?: User[];
     already_added_subgroups?: UserGroup[];
     ignored_deactivated_groups?: UserGroup[];
 }): void {
+    const newly_added_user_count = newly_added_users?.length ?? 0;
```

Comment:
Can we return early if there is an error message? And then add asserts for the optional fields like `newly_added_users`, `newly_added_subgroups`, etc. so we do not have to use `?? 0` and `generate_members_added_success_message` parameter types will also not be optional.

Reply:
Or maybe we can have different function for errors and success. Would help in avoiding optional types and assert statements. And the function for error would anwyays be a small function.